core/editor/src/tool/tools/eyedropper.rs

```diff
@@ -16,27 +18,19 @@ pub enum EyedropperMessage {
 impl<'a> MessageHandler<ToolMessage, ToolActionHandlerData<'a>> for Eyedropper {
 	fn process_action(&mut self, action: ToolMessage, data: ToolActionHandlerData<'a>, responses: &mut VecDeque<Message>) {
 		let mouse_pos = data.2.mouse.position;
-		let (x, y) = (mouse_pos.x as f64, mouse_pos.y as f64);
-		let (point_1, point_2) = (
-			DVec2::new(x - SELECTION_TOLERANCE, y - SELECTION_TOLERANCE),
-			DVec2::new(x + SELECTION_TOLERANCE, y + SELECTION_TOLERANCE),
-		);
-
-		let quad = [
-			DVec2::new(point_1.x, point_1.y),
-			DVec2::new(point_2.x, point_1.y),
-			DVec2::new(point_2.x, point_2.y),
-			DVec2::new(point_1.x, point_2.y),
-		];
+		let tolerance = DVec2::splat(SELECTION_TOLERANCE);
+		let quad = Quad::from_box([mouse_pos.as_f64() - tolerance, mouse_pos.as_f64() + tolerance]);
 
 		if let Some(path) = data.0.document.intersects_quad_root(quad).last() {
 			if let Ok(layer) = data.0.document.layer(path) {
-				if let Some(color) = layer.style.fill().and_then(|fill| fill.color()) {
-					match action {
-						ToolMessage::Eyedropper(EyedropperMessage::LeftMouseDown) => responses.push_back(ToolMessage::SelectPrimaryColor(color).into()),
-						ToolMessage::Eyedropper(EyedropperMessage::RightMouseDown) => responses.push_back(ToolMessage::SelectSecondaryColor(color).into()),
-						_ => {}
-					}
+				if let LayerDataType::Shape(s) = &layer.data {
```

Comment:
I'm not a fan of replacing `and_then` with nested maps - `and_then` is designed to achieve exactly this.